File: westcoast/src/main/java/frc/commands/targeting/ConstantlyAlignToTarget.java
package frc.commands.targeting;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.subsystems.DriveTrainSubsystem;
import org.jetbrains.annotations.NotNull;

import static frc.utils.Maths.clamp;

public final class ConstantlyAlignToTarget extends CommandBase {
    @NotNull
    private final PIDController turn;
    @NotNull
    private final PIDController forward;
    @NotNull
    private final DriveTrainSubsystem drive;

    public ConstantlyAlignToTarget(
            @NotNull PIDController turn,
            @NotNull PIDController forward,
            @NotNull DriveTrainSubsystem drive
    ) {
        this.turn = turn;
        this.forward = forward;
        this.drive = drive;
    }

    @Override
    public void initialize() {
        turn.reset();
        forward.reset();
    }

    @Override
    public void execute() {
        drive.arcade(
                clamp(-forward.calculate(drive.vision.getDistance(), 5.0), 0.4),
                clamp(turn.calculate(drive.vision.getAlignX(), 0.0), 0.4),
                false
        );
    }

    @Override
    public void end(boolean interrupted) {
        drive.stop();
    }
}

File: westcoast/src/main/kotlin/frc/commands/targeting/ConstantlyAlignToTarget.kt
package frc.commands.targeting

import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.wpilibj2.command.CommandBase
import frc.subsystems.DriveTrainSubsystem

class ConstantlyAlignToTarget(
    private val turn: PIDController,
    private val forward: PIDController,
    private val drive: DriveTrainSubsystem
) : CommandBase() {
    override fun initialize() {
        turn.reset()
        forward.reset()
    }

    override fun execute() {
        drive.arcade(
            -forward.calculate(drive.vision.distance, 5.0).coerceIn(-0.4, 0.4),
            turn.calculate(drive.vision.xOffset, 0.0).coerceIn(-0.4, 0.4),
            false
        )
    }

    override fun end(interrupted: Boolean) {
        drive.stop()
    }
}